Flesh out The Workload Identity Token section (#8)
yaronf/wimse-s2s: draft-sheffer-wimse-s2s-protocol.md

# Application Level Service To Service Authentication {#app-level}

## The Workload Identity Token {#to-wit}

The Workload Identity Token (WIT) is a JWS {{RFC7515}} signed JWT {{RFC7519}} that represents the identity of a workload.
It is issued by the Identity Server and binds a public key to the workload identity.
A WIT MUST contain the following:

* in the JOSE header:
    * `alg`: An identifier for a JWS asymmetric digital signature algorithm
     (registered algorithm identifiers are listed in the IANA JOSE Algorithms registry {{IANA.JOSE.ALGS}}). The value `none` MUST NOT be used.
    * `typ`: the WIT is explicitly typed, as recommended in {{Section 3.11 of RFC8725}}, using the `wimse-id+jwt` media type.
* in the JWT claims:
    * `iss`: The issuer of the token, which is the Identity Server, represented by a URI.

>     * `sub`: The subject of the token, which is the identity of the workload, represented by a URI.

Author: https://github.com/yaronf/wimse-s2s/pull/7/files#r1629377450 could help

>     * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).

Author: I am one of those people (and recently have been watching a confused, confusing, and painful conversation in a different WG of a different SDO play out around some somewhat arbitrary but explicit time windows) but something general and not normative like the text you suggested seems like a reasonable addition.

>     * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).

Author: @yaronf's suggested suggestion
```suggestion
    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
```

    * `jti`: A unique identifier for the token.
    * `cnf`: A confirmation claim containing the public key of the workload using the `jwk` member as defined in {{Section 3.2 of RFC7800}}.

> 

Author: say something here ish about needing to prove possession of the cnf.jwk key in order to use/present the WIT...

Author: https://github.com/yaronf/wimse-s2s/pull/8/files#r1630317742

An example WIT might look like this (all examples, of course, are non-normative and with line breaks and extra space for readability):

~~~ jwt
eyJ0eXAiOiJ3aW1zZS1pZCtqd3QiLCJhbGciOiJFUzI1NiIsImtpZCI6Ikp1bmUgNSJ9.
eyJpc3MiOiJ3aW1zZTovL2V4YW1wbGUuY29tL3RydXN0ZWQtY2VudHJhbC1hdXRob3Jpd
HkiLCJleHAiOjE3MTc2MTI0NzAsInN1YiI6IndpbXNlOi8vZXhhbXBsZS5jb20vc3BlY2
lmaWMtd29ya2xvYWQiLCJqdGkiOiJ4LV8xQ1RMMmNjYTNDU0U0Y3diX18iLCJjbmYiOns
iandrIjp7Imt0eSI6Ik9LUCIsImNydiI6IkVkMjU1MTkiLCJ4IjoiX2FtUkMzWXJZYkho
SDFSdFlyTDhjU21URE1oWXRPVVRHNzhjR1RSNWV6ayJ9fX0.rOSUMR8I5WhM5C704l3iV
dY0zFqxhugJ8Jo2xo39G7FqUTbwTzAGdpz2lHp6eL1M486XmRgl3uyjj6R_iuzNOA
~~~

The decoded JOSE header of the WIT from the example above is shown here:

~~~ json
{
 "typ": "wimse-id+jwt",
 "alg": "ES256",

>  "kid": "June 5"

Author: in the JWK itself plain random strings in examples start to look like maybe just another part of the key. 
https://github.com/yaronf/wimse-s2s/pull/8/files#diff-99cedafce05f1e1fafafbc415a81b34da2df6b52ad13892735fbdab5c43abf08R214 
so I arbitrarily chose something that I thought wouldn't blend in as much. But I'm not married to it.

}
~~~

The decoded JWT claims of the WIT from the example above are shown here:

~~~ json
{
 "iss": "wimse://example.com/trusted-central-authority",
 "exp": 1717612470,
 "sub": "wimse://example.com/specific-workload",
 "jti": "x-_1CTL2cca3CSE4cwb__",
 "cnf": {

>   "jwk": {

Reviewer: This key would shortly be used to sign its own JWTs (or HTTP messages). Doesn't it make sense to include a "kid" for it, too?

Author: Thought about that but (without having gone deep into the details) didn't see the need for a kid here as the relationship will be implied and the key will be there in the WIT.

Reviewer: As we discussed, `jwk` allows you to include a `kid`. So I'm fine with not mentioning it.

# Application Level Service To Service Authentication {#app-level}

## The Workload Identity Token {#to-wit}

The Workload Identity Token (WIT) is a JWS {{RFC7515}} signed JWT {{RFC7519}} that represents the identity of a workload.
It is issued by the Identity Server and binds a public key to the workload identity.
A WIT MUST contain the following:

* in the JOSE header:
    * `alg`: An identifier for a JWS asymmetric digital signature algorithm
     (registered algorithm identifiers are listed in the IANA JOSE Algorithms registry {{IANA.JOSE.ALGS}}). The value `none` MUST NOT be used.
    * `typ`: the WIT is explicitly typed, as recommended in {{Section 3.11 of RFC8725}}, using the `wimse-id+jwt` media type.
* in the JWT claims:
    * `iss`: The issuer of the token, which is the Identity Server, represented by a URI.
    * `sub`: The subject of the token, which is the identity of the workload, represented by a URI.
    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
    * `jti`: A unique identifier for the token.

>     * `cnf`: A confirmation claim containing the public key of the workload using the `jwk` member as defined in {{Section 3.2 of RFC7800}}.

Author: my own suggestion from https://github.com/yaronf/wimse-s2s/pull/8/files#r1629359149
```suggestion
    * `cnf`: A confirmation claim containing the public key of the workload using the `jwk` member as defined in {{Section 3.2 of RFC7800}}. 
     The workload MUST prove possession of the corresponding private key when presenting the WIT to another party. 
```

The Workload Identity Token (WIT) is a JWS {{RFC7515}} signed JWT {{RFC7519}} that represents the identity of a workload.
It is issued by the Identity Server and binds a public key to the workload identity.
A WIT MUST contain the following:

* in the JOSE header:
    * `alg`: An identifier for a JWS asymmetric digital signature algorithm
     (registered algorithm identifiers are listed in the IANA JOSE Algorithms registry {{IANA.JOSE.ALGS}}). The value `none` MUST NOT be used.
    * `typ`: the WIT is explicitly typed, as recommended in {{Section 3.11 of RFC8725}}, using the `wimse-id+jwt` media type.
* in the JWT claims:
    * `iss`: The issuer of the token, which is the Identity Server, represented by a URI.
    * `sub`: The subject of the token, which is the identity of the workload, represented by a URI.
    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
    * `jti`: A unique identifier for the token.
    * `cnf`: A confirmation claim containing the public key of the workload using the `jwk` member as defined in {{Section 3.2 of RFC7800}}.

>      The workload MUST prove possession of the corresponding private key when presenting the WIT to another party.

Reviewer: I think this statement is not useful because most developers don't know what it means and whether they're even compliant with this MUST. Moreover, the next 3 sections will each include a (sometimes implicit) proof of possession.

Author: I'd like to keep it to call attention to the need to do something other than just accept this as a bearer token. RFC7800 is not super well known and is a little opaque. And not even clear (to me anyway) if the presence of the cnf requires PoP. 

What about mentioning or referencing the next sections to give context about how one would be compliant with this MUST?

Reviewer: Yes, please refer to the next sections. But maybe also reverse the sense of the MUST: the WIT MUST NOT be used as a bearer token. Instead, it should be used in conjunction with one of the methods in sections foo, bar, and baz.

Author: will do

# Application Level Service To Service Authentication {#app-level}

## The Workload Identity Token {#to-wit}

> The Workload Identity Token (WIT) is a JWS {{RFC7515}} signed JWT {{RFC7519}} that represents the identity of a workload.

Author: "let's ignore @arndt-s" - @yaronf  :)

It is issued by the Identity Server and binds a public key to the workload identity.

> A WIT MUST contain the following:

Reviewer 2: +1 for WIT

* in the JOSE header:
    * `alg`: An identifier for a JWS asymmetric digital signature algorithm
     (registered algorithm identifiers are listed in the IANA JOSE Algorithms registry {{IANA.JOSE.ALGS}}). The value `none` MUST NOT be used.
    * `typ`: the WIT is explicitly typed, as recommended in {{Section 3.11 of RFC8725}}, using the `wimse-id+jwt` media type.
* in the JWT claims:
    * `iss`: The issuer of the token, which is the Identity Server, represented by a URI.
    * `sub`: The subject of the token, which is the identity of the workload, represented by a URI.
    * `exp`: The expiration time of the token (as defined in {{Section 4.1.4 of RFC7519}}).
      WITs should be refreshed regularly, e.g. on the order of hours.
    * `jti`: A unique identifier for the token.
    * `cnf`: A confirmation claim containing the public key of the workload using the `jwk` member as defined in {{Section 3.2 of RFC7800}}.

>      The WIT MUST NOT be used as a bearer token. Instead, it should be used in conjunction with one of the methods in {{dpop-esque-auth}} or {{http-sig-auth}}.

Reviewer 3: ```suggestion
The WIT MUST NOT be used as a single bearer token.
```
Otherwise it sounds like it should not be put into the Authorization header at all.

Reviewer 3: Discussed offline.

Reviewer 3: ```suggestion
WITs should always be used in conjunction with {{dpop-esque-auth}} or {{http-sig-auth}}. For this reason it MUST NOT be used as a bearer token and not go into the `Authorization` header.
```

Author: thanks @arndt-s - using some of your wording there and some of my old wording mixed together to hopefully not be too terrible. But we can revisit later, if need be. 
```suggestion
     The workload MUST prove possession of the corresponding private key when presenting the WIT to another party, which can be accomplished by using it in conjunction with one of the methods in {{dpop-esque-auth}} or {{http-sig-auth}}. As such, it MUST NOT be used as a bearer token and is not intended for use in the `Authorization` header.
```

 "crv":"P-256"
}
~~~
{: title="Example Identity Server Key"}

### The WIT HTTP Header {#wit-http-header}

A WIT is conveyed in an HTTP header field named `Workload-Identity-Token`.

For those who celebrate, ABNF {{RFC5234}} for the value of `Workload-Identity-Token` header field is provided in {{wit-header-abnf}}:

~~~ abnf
ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
DIGIT = %x30-39 ; 0-9
base64url = 1*(ALPHA / DIGIT / "-" / "_")
WIT =  base64url "." base64url "." base64url
~~~~
{: #wit-header-abnf title="Workload-Identity-Token Header Field ABNF"}

The following shows the WIT from the {{example-wit}} in an example of a `Workload-Identity-Token` header field:

~~~ http-message
Workload-Identity-Token: eyJ0eXAiOiJ3aW1zZS1pZCtqd3QiLCJhbGciOiJFUzI1
 NiIsImtpZCI6Ikp1bmUgNSJ9.eyJpc3MiOiJ3aW1zZTovL2V4YW1wbGUuY29tL3RydXN
 0ZWQtY2VudHJhbC1hdXRob3JpdHkiLCJleHAiOjE3MTc2MTI0NzAsInN1YiI6IndpbXN
 lOi8vZXhhbXBsZS5jb20vc3BlY2lmaWMtd29ya2xvYWQiLCJqdGkiOiJ4LV8xQ1RMMmN
 jYTNDU0U0Y3diX18iLCJjbmYiOnsiandrIjp7Imt0eSI6Ik9LUCIsImNydiI6IkVkMjU
 1MTkiLCJ4IjoiX2FtUkMzWXJZYkhoSDFSdFlyTDhjU21URE1oWXRPVVRHNzhjR1RSNWV
 6ayJ9fX0.rOSUMR8I5WhM5C704l3iVdY0zFqxhugJ8Jo2xo39G7FqUTbwTzAGdpz2lHp
6eL1M486XmRgl3uyjj6R_iuzNOA

> ~~~

Reviewer: If you want it to be a modern+shiny structured field value (RFC 8941), I think you need colons around the base64 text.

Reviewer: OK, this is not SFV. Sad.
